File: backend/src/services/mcp/toolRegistry.ts
/**
 * ToolRegistry - 统一工具注册表 + 工具缓存 + mtime 过期检测
 *
 * 管理本地 brainTools 和外部 MCP 工具的合并列表
 * 提供 getAllTools() 供大脑 OODA 循环使用
 *
 * Requirements: 6.1, 6.2, 6.3, 6.4, 6.5, 14.1, 14.2, 14.3, 14.4, 14.5
 */

import { Mutex } from 'async-mutex';
import * as fs from 'fs/promises';
import { logger } from '../../utils/logger';
import { auditLogger } from '../ai-ops/auditLogger';

// ─── Types ───────────────────────────────────────────────────────────────────

/** AgentTool 兼容接口（与 mastraAgent.ts 一致） */
export interface AgentToolCompat {
  name: string;
  description: string;
  parameters: Record<string, { type: string; description: string; required?: boolean }>;
  execute: (params: Record<string, unknown>) => Promise<unknown>;
}

/** 统一工具接口 */
export interface UnifiedTool {
  name: string;
  description: string;
  parameters: Record<string, any>;
  source: 'local' | 'mcp';
  serverId?: string;
  healthy: boolean;
  execute: (params: Record<string, unknown>) => Promise<unknown>;
}

/** MCP 工具定义（从外部 Server 发现） */
export interface McpToolDefinition {
  name: string;
  description?: string;
  inputSchema?: Record<string, any>;
}

/** 工具缓存条目 */
interface ToolCacheEntry {
  tools: UnifiedTool[];
  cachedAt: number;
  configMtime: number;
}

/** McpClientManager 前向引用接口 */
export interface IToolForwarder {
  forwardCall(serverId: string, toolName: string, args: Record<string, unknown>): Promise<unknown>;
}

// ─── ToolRegistry ────────────────────────────────────────────────────────────

/**
 * 🔴 FIX 1.10: 将 MCP JSON Schema 格式的 inputSchema 归一化为扁平的 ToolParam 格式
 * MCP 工具的 inputSchema 是标准 JSON Schema（{ type: "object", properties: {...}, required: [...] }）
 * 而 Brain 的 brainTools 使用扁平格式（{ paramName: { type, description, required } }）
 * 此函数将前者转换为后者，确保 ReAct 循环能统一解析参数
 */
export function normalizeInputSchema(inputSchema: Record<string, any>): Record<string, { type: string; description: string; required?: boolean }> {
  if (!inputSchema || typeof inputSchema !== 'object') return {};

  // 如果已经是扁平格式（没有 properties 字段，且值都有 type），直接返回
  if (!inputSchema.properties && !inputSchema.type) {
    // 可能已经是扁平格式，检查第一个值
    const firstVal = Object.values(inputSchema)[0];
    if (firstVal && typeof firstVal === 'object' && 'type' in (firstVal as any)) {
      return inputSchema as Record<string, { type: string; description: string; required?: boolean }>;
    }
    return {};
  }

  const properties = inputSchema.properties;
  if (!properties || typeof properties !== 'object') return {};

  const requiredFields = new Set<string>(
    Array.isArray(inputSchema.required) ? inputSchema.required : []
  );

  const result: Record<string, { type: string; description: string; required?: boolean }> = {};

  for (const [key, schema] of Object.entries(properties)) {
    const prop = schema as Record<string, any>;
    result[key] = {
      type: typeof prop.type === 'string' ? prop.type : 'string',
      description: typeof prop.description === 'string' ? prop.description : '',
      required: requiredFields.has(key),
    };
  }

  return result;
}

export class ToolRegistry {
  private localTools: Map<string, UnifiedTool> = new Map();
  private externalTools: Map<string, UnifiedTool> = new Map();
  private cache: ToolCacheEntry | null = null;
  private cacheLock: Mutex = new Mutex();
  private configFilePath: string;
  private toolForwarder: IToolForwarder | null = null;
  /** 外部回调：缓存失效时触发重新发现 */
  private onCacheInvalidated: (() => Promise<void>) | null = null;

  constructor(configFilePath: string) {
    this.configFilePath = configFilePath;
  }

  /**
   * 设置工具转发器（McpClientManager 注入）
   */
  setToolForwarder(forwarder: IToolForwarder): void {
    this.toolForwarder = forwarder;
  }

  /**
   * 设置缓存失效回调（McpClientManager 注入）
   */
  setOnCacheInvalidated(callback: () => Promise<void>): void {
    this.onCacheInvalidated = callback;
  }

  /**
   * 注册本地 brainTools
   */
  registerLocalTools(brainTools: AgentToolCompat[]): void {
    for (const tool of brainTools) {
      const unified: UnifiedTool = {
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters,
        source: 'local',
        healthy: true,
        execute: tool.execute,
      };
      this.localTools.set(tool.name, unified);
    }
    logger.info(`[ToolRegistry] Registered ${brainTools.length} local tools`);
  }

  /**
   * 注册外部 MCP 工具（名称格式 mcp:{serverId}:{toolName}）
   */
  registerExternalTools(serverId: string, tools: McpToolDefinition[]): void {
    for (const tool of tools) {
      const unifiedName = `mcp:${serverId}:${tool.name}`;
      const unified: UnifiedTool = {
        name: unifiedName,
        description: tool.description || '',
        parameters: tool.inputSchema || {},
        source: 'mcp',
        serverId,
        healthy: true,
        execute: async (params) => {
          // 🔴 FIX 1.10: 调用前检查 MCP 服务器连接健康状态
          const currentTool = this.externalTools.get(unifiedName);
          if (currentTool && !currentTool.healthy) {
            return {
              success: false,
              error: `MCP server "${serverId}" disconnected`,
              _brainHint: `此 MCP 工具 (${unifiedName}) 不可用，对应的 MCP 服务器已断连。请尝试本地替代工具或跳过此操作。`,
            };
          }
          if (!this.toolForwarder) {
            throw new Error(`No tool forwarder set. Cannot call external tool: ${unifiedName}`);
          }
          // 🔴 FIX 1.10e: 捕获 forwardCall 异常，返回带 _brainHint 的结构化错误
          try {
            return await this.toolForwarder.forwardCall(serverId, tool.name, params);
          } catch (err) {
            const errMsg = err instanceof Error ? err.message : String(err);
            // 标记服务器为不健康，后续调用直接走健康检查拦截
            this.setServerHealth(serverId, false);
            return {
              success: false,
              error: `MCP tool "${unifiedName}" execution failed: ${errMsg}`,
              _brainHint: `MCP 工具 (${unifiedName}) 调用失败 (${errMsg})。该 MCP 服务器已标记为不可用。请尝试功能等价的本地工具或跳过此操作。`,
            };
          }
        },
      };
      this.externalTools.set(unifiedName, unified);
    }
    // 清除缓存，下次 getAllTools 会重建
    this.cache = null;
    logger.info(`[ToolRegistry] Registered ${tools.length} external tools from server: ${serverId}`);
  }

  /**
   * 注销外部工具（按 serverId）
   */
  unregisterExternalTools(serverId: string): void {
    let removed = 0;
    for (const [name, tool] of this.externalTools) {
      if (tool.serverId === serverId) {
        this.externalTools.delete(name);
        removed++;
      }
    }
    this.cache = null;
    logger.info(`[ToolRegistry] Unregistered ${removed} external tools from server: ${serverId}`);
  }

  /**
   * 标记外部 Server 健康状态
   */
  setServerHealth(serverId: string, healthy: boolean): void {
    for (const tool of this.externalTools.values()) {
      if (tool.serverId === serverId) {
        tool.healthy = healthy;
      }
    }
    // 健康状态变化时清除缓存
    this.cache = null;
  }

  /**
   * 获取全量工具列表（本地 + 健康的外部工具）
   * 内部先检查 mtime 是否过期
   */
  async getAllTools(): Promise<UnifiedTool[]> {
    // 检查缓存 mtime 是否过期
    const isStale = await this.checkCacheStaleness();
    if (isStale) {
      await this.invalidateAndRefresh();
    }

    // 如果有有效缓存，直接返回
    if (this.cache) {
      return this.cache.tools;
    }

    // 构建工具列表
    const tools: UnifiedTool[] = [];

    // 本地工具全部加入
    for (const tool of this.localTools.values()) {
      tools.push(tool);
    }

    // 外部工具仅加入 healthy=true 的，并归一化 parameters
    for (const tool of this.externalTools.values()) {
      if (tool.healthy) {
        tools.push({
          ...tool,
          parameters: normalizeInputSchema(tool.parameters),
        });
      }
    }

    // 缓存结果
    const mtime = await this.getConfigMtime();
    this.cache = {
      tools,
      cachedAt: Date.now(),
      configMtime: mtime,
    };

    return tools;
  }

  /**
   * 执行工具（自动路由本地/外部）
   */
  async executeTool(toolName: string, params: Record<string, unknown>): Promise<unknown> {
    // 优先查找本地工具
    const localTool = this.localTools.get(toolName);
    if (localTool) {
      return localTool.execute(params);
    }

    // 查找外部工具
    const externalTool = this.externalTools.get(toolName);
    if (externalTool) {
      if (!externalTool.healthy) {
        throw new Error(`External tool ${toolName} is unhealthy`);
      }
      const startTime = Date.now();
      try {
        const result = await externalTool.execute(params);
        const duration = Date.now() - startTime;
        // 审计日志
        auditLogger.log({
          action: 'mcp_external_tool_call' as any,
          actor: 'system',
          source: 'mcp_client',
          details: {
            toolName,
            serverId: externalTool.serverId,
            inputParams: params,
            result: typeof result === 'object' ? JSON.stringify(result).slice(0, 500) : String(result).slice(0, 500),
            durationMs: duration,
            success: true,
          },
        }).catch(() => { /* non-critical */ });
        return result;
      } catch (error) {
        const duration = Date.now() - startTime;
        auditLogger.log({
          action: 'mcp_external_tool_call' as any,
          actor: 'system',
          source: 'mcp_client',
          details: {
            toolName,
            serverId: externalTool.serverId,
            inputParams: params,
            error: error instanceof Error ? error.message : String(error),
            durationMs: duration,
            success: false,
          },
        }).catch(() => { /* non-critical */ });
        throw error;
      }
    }

    throw new Error(`Tool not found: ${toolName}`);
  }

  /**
   * 检查缓存是否过期（配置文件 mtime 变化）
   */
  private async checkCacheStaleness(): Promise<boolean> {
    if (!this.cache) return false; // 无缓存不算过期

    try {
      const currentMtime = await this.getConfigMtime();
      return currentMtime > this.cache.configMtime;
    } catch {
      // 无法读取 mtime，跳过检测
      return false;
    }
  }

  /**
   * 缓存失效并触发重新发现（通过 cacheLock 串行化）
   */
  private async invalidateAndRefresh(): Promise<void> {
    await this.cacheLock.runExclusive(async () => {
      // double-check：锁内再次检查是否仍然过期
      const stillStale = await this.checkCacheStaleness();
      if (!stillStale && this.cache) return;

      logger.info('[ToolRegistry] Cache invalidated due to config mtime change. Triggering rediscovery...');
      this.cache = null;

      if (this.onCacheInvalidated) {
        try {
          await this.onCacheInvalidated();
        } catch (err) {
          logger.warn(`[ToolRegistry] Cache invalidation callback failed: ${err instanceof Error ? err.message : String(err)}`);
        }
      }
    });
  }

  /**
   * 读取配置文件 mtime
   */
  private async getConfigMtime(): Promise<number> {
    try {
      const stat = await fs.stat(this.configFilePath);
      return stat.mtimeMs;
    } catch {
      return 0;
    }
  }

  /**
   * 获取已注册的外部工具数量（按 serverId）
   */
  getExternalToolCount(serverId: string): number {
    let count = 0;
    for (const tool of this.externalTools.values()) {
      if (tool.serverId === serverId) count++;
    }
    return count;
  }

  /**
   * 获取所有外部工具（按 serverId）
   */
  getExternalToolsByServer(serverId: string): UnifiedTool[] {
    const tools: UnifiedTool[] = [];
    for (const tool of this.externalTools.values()) {
      if (tool.serverId === serverId) tools.push(tool);
    }
    return tools;
  }
}

File: opsevo-python/src/opsevo/services/mcp/tool_registry.py
"""
ToolRegistry — 统一工具注册表

合并本地 brainTools 和远程 MCP 工具到统一接口。
支持工具缓存、健康状态跟踪、工具转发、回调通知。
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Callable, Protocol

import structlog

logger = structlog.get_logger(__name__)


@dataclass
class UnifiedTool:
    """统一工具定义。"""

    name: str
    description: str
    input_schema: dict[str, Any] = field(default_factory=dict)
    source: str = "local"  # local | external
    server_id: str | None = None
    healthy: bool = True


class IToolForwarder(Protocol):
    """工具转发器协议。"""

    async def forward_call(
        self, server_id: str, tool_name: str, args: dict[str, Any]
    ) -> Any: ...


class ToolRegistry:
    """统一工具注册表。"""

    def __init__(self, config_file_path: str = "") -> None:
        self._config_file_path = config_file_path
        self._local_tools: dict[str, UnifiedTool] = {}
        self._external_tools: dict[str, dict[str, UnifiedTool]] = {}  # server_id -> {name: tool}
        self._server_health: dict[str, bool] = {}
        self._forwarder: IToolForwarder | None = None
        self._local_handlers: dict[str, Any] = {}
        self._cache_timestamp: float = 0.0
        self._on_cache_invalidated: Any = None
        self._on_tools_changed_callbacks: list[Callable[[], None]] = []

    def set_tool_forwarder(self, forwarder: IToolForwarder) -> None:
        self._forwarder = forwarder

    def set_on_cache_invalidated(self, callback: Any) -> None:
        self._on_cache_invalidated = callback

    def register_on_tools_changed(self, callback: Callable[[], None]) -> None:
        """Register a callback to be notified when tools change."""
        self._on_tools_changed_callbacks.append(callback)

    def _notify_tools_changed(self) -> None:
        """Invoke all registered on_tools_changed callbacks."""
        for cb in self._on_tools_changed_callbacks:
            try:
                cb()
            except Exception:
                logger.warning("on_tools_changed callback failed")

    # ------------------------------------------------------------------
    # 本地工具注册
    # ------------------------------------------------------------------

    def register_local_tools(
        self, tools: list[dict[str, Any]]
    ) -> None:
        """注册本地 brainTools。"""
        for t in tools:
            name = t.get("name", "")
            tool = UnifiedTool(
                name=name,
                description=t.get("description", ""),
                input_schema=t.get("inputSchema", t.get("input_schema", {})),
                source="local",
            )
            self._local_tools[name] = tool
            if "handler" in t:
                self._local_handlers[name] = t["handler"]
        self._cache_timestamp = time.time()
        logger.info("Local tools registered", count=len(tools))

    # ------------------------------------------------------------------
    # 外部工具注册
    # ------------------------------------------------------------------

    def register_external_tools(
        self, server_id: str, tools: list[Any]
    ) -> None:
        """注册外部 MCP Server 发现的工具。"""
        tool_map: dict[str, UnifiedTool] = {}
        for t in tools:
            name = getattr(t, "name", "") if hasattr(t, "name") else t.get("name", "")
            desc = getattr(t, "description", "") if hasattr(t, "description") else t.get("description", "")
            schema = getattr(t, "input_schema", {}) if hasattr(t, "input_schema") else t.get("inputSchema", {})
            tool_map[name] = UnifiedTool(
                name=name,
                description=desc,
                input_schema=schema,
                source="external",
                server_id=server_id,
            )
        self._external_tools[server_id] = tool_map
        self._server_health[server_id] = True
        self._cache_timestamp = time.time()
        logger.info("External tools registered", server_id=server_id, count=len(tools))
        self._notify_tools_changed()

    def unregister_external_tools(self, server_id: str) -> None:
        self._external_tools.pop(server_id, None)
        self._server_health.pop(server_id, None)
        self._cache_timestamp = time.time()
        self._notify_tools_changed()

    def set_server_health(self, server_id: str, healthy: bool) -> None:
        prev = self._server_health.get(server_id)
        self._server_health[server_id] = healthy
        if prev != healthy:
            self._notify_tools_changed()

    # ------------------------------------------------------------------
    # 查询
    # ------------------------------------------------------------------

    def get_all_tools(self) -> list[UnifiedTool]:
        """获取所有可用工具（本地 + 健康的外部）。"""
        tools: list[UnifiedTool] = list(self._local_tools.values())
        for sid, tool_map in list(self._external_tools.items()):
            if self._server_health.get(sid, False):
                tools.extend(tool_map.values())
        return tools

    def get_external_tool_count(self, server_id: str) -> int:
        return len(self._external_tools.get(server_id, {}))

    def get_external_tools_by_server(self, server_id: str) -> list[UnifiedTool]:
        return list(self._external_tools.get(server_id, {}).values())

    def get_all_tool_definitions(self) -> list[dict[str, Any]]:
        """Return unified tool definitions in OpenAI schema format with source annotation.

        Deduplicates by name — local tools take priority over external.
        Only includes tools from healthy external servers.

        Requirements: 5.2, 5.3, 5.5, 5.6, 5.7
        """
        seen_names: set[str] = set()
        definitions: list[dict[str, Any]] = []

        # Local tools first (higher priority)
        for tool in self._local_tools.values():
            seen_names.add(tool.name)
            definitions.append(self._unified_tool_to_definition(tool))

        # External tools from healthy servers (dedup by name)
        for sid, tool_map in list(self._external_tools.items()):
            if not self._server_health.get(sid, False):
                continue
            for tool in tool_map.values():
                if tool.name not in seen_names:
                    seen_names.add(tool.name)
                    definitions.append(self._unified_tool_to_definition(tool))

        return definitions

    @staticmethod
    def _unified_tool_to_definition(tool: UnifiedTool) -> dict[str, Any]:
        """Convert a UnifiedTool to an OpenAI function calling schema dict with source."""
        desc = tool.description
        return {
            "type": "function",
            "function": {
                "name": tool.name,
                "description": desc,
                "parameters": tool.input_schema if tool.input_schema else {"type": "object", "properties": {}},
            },
            "source": tool.source,
            "server_id": tool.server_id,
        }

    # ------------------------------------------------------------------
    # 执行
    # ------------------------------------------------------------------

    async def execute_tool(
        self, tool_name: str, params: dict[str, Any]
    ) -> Any:
        """执行工具（本地直接调用，外部通过 forwarder 转发）。"""
        # 本地工具
        if tool_name in self._local_tools:
            handler = self._local_handlers.get(tool_name)
            if handler:
                return await handler(params)
            raise RuntimeError(f"No handler for local tool: {tool_name}")

        # 外部工具 (snapshot to avoid RuntimeError if dict mutates concurrently)
        for sid, tool_map in list(self._external_tools.items()):
            if tool_name in tool_map and self._server_health.get(sid, False):
                if not self._forwarder:
                    raise RuntimeError("No tool forwarder configured")
                return await self._forwarder.forward_call(sid, tool_name, params)

        raise RuntimeError(f"Tool not found: {tool_name}")
